components/plurals/src/lib.rs

```diff
@@ -550,3 +552,104 @@ impl PluralRules {
             .chain(Some(PluralCategory::Other))
     }
 }
+
+/// A struct which provides the ability to retrieve an appropriate [`Plural Category`] for a range
+/// from the categories of its endpoints.
+///
+/// # Examples
+///
+/// ```
+/// use icu::locid::locale;
+/// use icu::plurals::{PluralCategory, PluralOperands};
+/// use icu::plurals::{PluralRanges, PluralRuleType, PluralRules};
+/// use std::convert::TryFrom;
+///
+/// let pr =
+///     PluralRules::try_new(&locale!("ar").into(), PluralRuleType::Cardinal)
+///         .expect("locale should be present");
+/// let ranges = PluralRanges::try_new(&locale!("ar").into())
+///     .expect("locale should be present");
+///
+/// let operands =
+///     PluralOperands::try_from(1).expect("Failed to parse to operands.");
+/// let operands2: PluralOperands =
+///     "2.0".parse().expect("Failed to parse to operands.");
+/// let start = pr.category_for(operands);
+/// let end = pr.category_for(operands2);
+///
+/// assert_eq!(ranges.category_for_range(start, end), PluralCategory::Other);
```

Comment:
having to acquire `PluralCategory`s from `PluralRules` first seems cumbersome

Reply:
I think the type should not have multiple responsibilities unless we're going to do the wrapper-type stuff that TitlecaseMapper does (which imo we don't need).

It would be fine to have a convenience method that takes in a PluralRules though

Reply:
I can see the argument that PluralRanges is not useful without PluralRules so maybe we should do a composition like LocaleCanonicalizer or TitlecaseMapper

Reply:
Ok, if we either add a convenience method or compose with `PluralRules`, should the new method return a `PluralCategory` unconditionally or should it return `Result` and check that `start > end` and `start > 0, end > 0`?

Note that ECMA402 will have to use the raw API anyways, because it [checks `start == end`](https://tc39.es/ecma402/#sec-resolvepluralrange) and skips resolving the range in that case.

Reply:
The composed version would be something like

```rust
struct PluralRanges {
    rules: DataPayload<ErasedPluralRulesV1Marker>,
    ranges: DataPayload<PluralRangesV1Marker>,
}

impl PluralRanges {
    pub fn category_for(x: impl Into<PluralOperands>) -> PluralCategory { ... }

    pub fn category_for_range(x: impl Into<PluralOperands>, y: impl Into<PluralOperands>) -> PluralCategory { ... }

    pub fn resolve_range(x: PluralCategory, y: PluralCategory) -> PluralCategory { ... }
}
```

Reply:
Unless UTS 35 imposes a restriction on start < end, then I don't think we should return a Result. Note that ECMA-402 allows ranges in the reverse order.

Reply:
> The composed version would be something like
> 
> ```rust
> struct PluralRanges {
>     rules: DataPayload<ErasedPluralRulesV1Marker>,
>     ranges: DataPayload<PluralRangesV1Marker>,
> }
> 
> impl PluralRanges {
>     pub fn category_for(x: impl Into<PluralOperands>) -> PluralCategory { ... }
> 
>     pub fn category_for_range(x: impl Into<PluralOperands>, y: impl Into<PluralOperands>) -> PluralCategory { ... }
> 
>     pub fn resolve_range(x: PluralCategory, y: PluralCategory) -> PluralCategory { ... }
> }
> ```

This would be `PluralRules` instead, right?

Reply:
Hmm, if that's the case, then I'm leaning towards the composed version, since it avoids having to provide the locale two times, which could lead to errors.

Reply:
`PluralRules` remains the way it is, with only `ErasedPluralRulesV1Marker`, and then we get a new type `PluralRanges` (maybe there's a better name), which has the functionality of `PluralRules` but also includes the plural ranges.

Reply:
Wouldn't be better to have `PluralRules` be:
```
struct PluralRanges {
    rules: DataPayload<ErasedPluralRulesV1Marker>,
    ranges: Option<DataPayload<PluralRangesV1Marker>>,
}
```
Then, we always return `end` as the resolved category if `ranges.is_none()`? Though, I can see that it would be a big pitfall to have a new API that doesn't work if the data is not updated.

Reply:
Two reasons to introduce a new type:

1. We enforce at build time that the correct constructors have been invoked. It reduces the space for developer footguns.
2. We don't affect the size of the existing `PluralRules` object. The new DataPayload increases the stack size and code size of the `Drop` impl, which, while fairly small, impacts all users of `PluralRules`, even though the majority of users don't need the ranges addendum.

Reply:
Understandable! Then something like `PluralRulesExtended` for the new type? Or do we want to be more explicit about the functionality e.g. `PluralRulesRanges`?

Reply:
Yep! I like `PluralRulesRanges` or `PluralRulesWithRanges`, then if there's another plural rules extension in the future, we don't clash the namespace.

Reply:
If we're not aligned on the API this should land behind an experimental feature.

Reply:
I am comfortable letting this land and be stabilized as-is. I would _prefer_ we used convenience APIs instead of making a combined type. This is not a position I hold strongly.

Reply:
Basically, if the two of you like this API as is I have no problem with it landing. However, if after reading my comment you change your mind, we should discuss it. Either way I am okay with having this land first and sorting that out later.

Reply:
Follow-up: https://github.com/unicode-org/icu4x/issues/4140